nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+package nameserver
+
+import (
+	"errors"
+	"github.com/miekg/dns"
+	"math"
+	"sort"
+	"sync"
+	"time"
+	//"fmt"
+	"math/rand"
+)
+
+var (
+	errInvalidCapacity = errors.New("Invalid cache capacity")
+	errCouldNotResolve = errors.New("Could not resolve")
+	errTimeout         = errors.New("Timeout while waiting for resolution")
+)
+
+const (
+	defPendingTimeout int = 5 // timeout for a resolution
+)
+
+type entryStatus uint8
+
+const (
+	stPending  entryStatus = iota // someone is waiting for the resolution
+	stResolved entryStatus = iota // resolved
+)
+
+const (
+	CacheLocalReply uint8 = 1 << iota // the reply was obtained from a local resolution
```

Comment:
The constant v boolean arguments argument is a good point (especially if you can make them varargs) -- I hadn't considered that.